### DISCUSSION TITLE
How to pass an n-dimentional numpy array from python -> c++

### DISCUSSION THREAD

**Discussion by munkybutt** [pybind/pybind11#3438] [Q&A] [answered]
I have a numpy array of numpy arrays of floats that I wish to pass to a c++ function that will read and modify the data as if it were a std::vector.
I am struggling to figure out how to do this.

What would the c++ argument type be for:
`np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]], dtype=float)`

This type declaration:
`py::array_t<py::array_t<float>> `
produces this error:
`Attempt to use a non-POD or unimplemented POD type as a numpy dtype`

I can't see to get this type declaration to work:
`py:array`

I have searched through documentation, github issues and examples I can find online but have not found anything that works or that I can make sense of, if anyone could help me with this, it would be much appreciated :)

**Comment by jiwaszki** [comment]
Can you add some reproducer? It seems quite unusual to encounter such error since pybind has a great basic out-of-the-box support for numpy.

Did you try using just replacing dtype with `dtype=np.float32` and py::array with `py::array_t<float>`? 
The  thing is that dimensions are later provided inside buffer info. There is NO need to create arrays like vectors in C++. And that is probably a problem that throws an error as you try to template array with some "custom type" (in your case `py::array<float>`).

**Comment by munkybutt** [comment, marked as answer]
Thanks for the reply - sorry for the delayed response.
I got an answer on stack overflow which works, effectively using Eigen Matrices:
https://stackoverflow.com/questions/69831285/pybind11-how-to-pass-an-n-dimentional-numpy-array-from-python-c
The conversion is automatic and using the Eigen::Ref, there is no copying of data
